src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs

```diff
@@ -0,0 +1,134 @@
+        private bool _isAnAuthenticationPlugin = true;
+
+        /// <summary>
+        /// Create a credential provider based on provided plugin
+        /// </summary>
+        /// <param name="pluginDiscoveryResult"></param>
+        /// <param name="logger"></param>
+        /// <exception cref="ArgumentNullException">if <paramref name="pluginDiscoveryResult"/> is null</exception>
+        /// <exception cref="ArgumentNullException">if <paramref name="logger"/> is null</exception>
+        /// <exception cref="ArgumentNullException">if <paramref name="pluginManager"/> is null</exception>
+        /// <exception cref="ArgumentException">if plugin file is not valid</exception>
+        public SecurePluginCredentialProvider(PluginManager pluginManager, PluginDiscoveryResult pluginDiscoveryResult, Common.ILogger logger)
+        {
+            if (pluginDiscoveryResult == null)
+            {
+                throw new ArgumentNullException(nameof(pluginDiscoveryResult));
+            }
+            if (pluginDiscoveryResult.PluginFile.State != PluginFileState.Valid)
+            {
+                throw new ArgumentException(string.Format(Resources.SecureCredentialProvider_InvalidPluginFile, pluginDiscoveryResult.PluginFile.State, pluginDiscoveryResult.Message), nameof(pluginDiscoveryResult));
+            }
+            _pluginManager = pluginManager ?? throw new ArgumentNullException(nameof(pluginManager));
+            _discoveredPlugin = pluginDiscoveryResult;
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            Id = $"{nameof(SecurePluginCredentialProvider)}_{pluginDiscoveryResult.PluginFile.Path}";
+        }
+
+        /// <summary>
+        /// Unique identifier of this credential provider
+        /// </summary>
+        public string Id { get; }
+
+        /// <param name="uri">The uri of a web resource for which credentials are needed.</param>
+        /// <param name="proxy">Ignored.  Proxy information will not be passed to plugins.</param>
+        /// <param name="type">
+        /// The type of credential request that is being made. Note that this implementation of
+        /// <see cref="ICredentialProvider"/> does not support providing proxy credenitials and treats
+        /// all other types the same.
+        /// </param>
+        /// <param name="isRetry">If true, credentials were previously supplied by this
+        /// provider for the same uri.</param>
+        /// <param name="message">A message provided by NuGet to show to the user when prompting.</param>
+        /// <param name="nonInteractive">If true, the plugin must not prompt for credentials.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A credential object.</returns>
+        public async Task<CredentialResponse> GetAsync(Uri uri, IWebProxy proxy, CredentialRequestType type, string message, bool isRetry, bool nonInteractive, CancellationToken cancellationToken)
```

Comment:
It appears that the LogRequestHandler is not wired up so I can't log things in my plugin.  Is that expected?

https://github.com/NuGet/NuGet.Client/blob/870340673ba75a95407ab55ce1f46f57fe9bf25b/src/NuGet.Core/NuGet.Protocol/Resources/DownloadResourcePlugin.cs#L139

Reply:
@jeffkl 
It's done consciously right now. 
There's no logger setup for the credential plugin. 
The above approach is a little bit spotty, because there's no contract on when a plugin can log. 

I've created a task to follow up on this:
https://github.com/NuGet/Home/issues/6762

Reply:
Would it be possible to link MSBuildLogger with credential plugin? since that's already been passed to this provider.

Reply:
@jainaashish 
Yeah, I think that'll be possible. 
The logger is propagated from RestoreTask -> RestoreRunner -> CredentialProviderBuilder -> CredentialProvider.